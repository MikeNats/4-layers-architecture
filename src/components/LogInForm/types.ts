
export type Submit = (event: React.FormEvent<HTMLFormElement>) => void

export type ValidationErros = {
  email: boolean,
  password: boolean, 
  isFormValid:boolean,
}

export type Props = {
  submitForm: Submit,
  authStatus: Boolean
  validationErrors: ValidationErros,
  updateEmail: Function,
  updatePassword: Function,
}
