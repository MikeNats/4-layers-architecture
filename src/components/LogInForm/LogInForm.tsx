import React from 'react'
import Input from '../common/Input/Input'
import Button from '../common/Button/Button'
import { LogInFormProps } from './types'

const LogInForm = ({
  submitForm,
  authFailed,
  errorEmail,
  errorPassword,
  isFormValid,
  updateEmail,
  updatePassword
}: LogInFormProps) => (
  <form 
    className="comp-logiInForm"
    onSubmit={submitForm}>
      
    <fieldset>
      <Input 
        type="text" 
        name="email" 
        id="email"
        label="user name:"
        error={errorEmail}
        errorMessage="Mandatory field"
        placeholder="email"
        onChangeHandler={updateEmail}
        required />
    </fieldset>

    <fieldset>
      <Input //default value
        type="text"
        name="password" 
        id="password"
        label="password:"
        placeholder="password"
        error={errorPassword}
        errorMessage="Mandatory field"
        onChangeHandler={updatePassword}
        required />
    </fieldset>

    <p className={ `errorMessage ${authFailed ? '' : 'hide'}` }>Invalid credentials</p>
    
    <fieldset className="align-button-fieldset-center">
      <Button
        disabled={!isFormValid}>
        submit</Button> 
    </fieldset>
</form>)

export default LogInForm