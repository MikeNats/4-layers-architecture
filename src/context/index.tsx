import React from 'react';

export type AppContextType = {
  authenticated: boolean,
  theme: string, 
  setAuth: Function
}
export const initialAppContext: AppContextType= { 
  authenticated: false,
  setAuth: function() {this.authenticated = true},
  theme: 'dark'
}

export const AppContext = React.createContext<Partial<AppContextType>>({})