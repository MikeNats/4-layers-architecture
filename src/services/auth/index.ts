
export default async (userName: string, password: string): Promise<{userId: string}> =>
  new Promise((resolve, reject) => {
    console.log('mock loging', userName, password)
    setTimeout(() => {
      resolve({
        userId: '123',
      })
    }, 1500)
})
