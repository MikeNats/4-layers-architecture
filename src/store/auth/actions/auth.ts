import { Action } from 'redux'
import { AUTH_ACTION_TYPE } from './types'
import { AsyncActions } from '../../types'
import { AsyncFailActions } from '../../types'

export const authActions: AsyncActions = {
  'REQUEST': (): Action => ({
    type: AUTH_ACTION_TYPE.REQUEST
  }),
  'SUCCESS': (): Action => ({
    type:  AUTH_ACTION_TYPE.SUCCESS,
  }), 
  'FAIL': (errorCode: number): AsyncFailActions => ({
    type: AUTH_ACTION_TYPE.FAIL,
    errorCode
  })
}
