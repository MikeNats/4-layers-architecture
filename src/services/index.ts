import { Dispatch } from 'redux'
import { ActionsType } from '../actions/types'
import { RequestType } from './types'
import axios, { AxiosRequestConfig } from 'axios'
import { isMock } from './utils'


export default  (axiosRequestConfig: AxiosRequestConfig, requestConfig?: RequestType): any => 
  requestConfig && requestConfig.asyncActionName ?
    (dispatch: Dispatch<ActionsType>): Promise<ActionsType> =>  {
      const { asyncActionName, actionArgs} = requestConfig;
      dispatch(asyncActionName['REQUEST'](actionArgs));
      return axios(isMock(axiosRequestConfig))
        .then(response => dispatch(asyncActionName['SUCCESS'](response.data)))
        .catch(error => dispatch(asyncActionName['FAIL'](error.errorCode || 404)))
   
    }: axios(isMock(axiosRequestConfig))