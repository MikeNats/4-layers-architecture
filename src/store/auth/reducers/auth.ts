import { AuthReducer } from './types'
import { AUTH_ACTION_TYPE, AuthAsyncActions} from '../actions/types'

export const initIdenitytReduerState: AuthReducer =  {
  isFetching: false,
  didInvalidate: false,
  authenticated: false,
  errorCode: null  
} 
    
export const authReducer = (state: AuthReducer = initIdenitytReduerState, action: AuthAsyncActions): AuthReducer  => {
  switch (action.type) {
    case AUTH_ACTION_TYPE.REQUEST:
      return {
        errorCode: null,
        authenticated: false,
        isFetching: true,
        didInvalidate: false,
      }
    case AUTH_ACTION_TYPE.SUCCESS:
      return {
        isFetching: false,  
        didInvalidate: false,
        authenticated: true,
        errorCode: null 
      }
    case AUTH_ACTION_TYPE.FAIL:
      return {
        authenticated: false,
        isFetching: false,
        didInvalidate: true,
        errorCode: action.errorCode
      }
    default: 
      return state
    }
  }