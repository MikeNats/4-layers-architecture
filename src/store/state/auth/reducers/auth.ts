import { LogInReducer } from './types'
import { AUTH_ACTION_TYPES, AuthAsyncActions} from '../actions/types'

export const initIdenitytReduerState: LogInReducer =  {
  isFetching: false,
  didInvalidate: false,
  authenticated: false,
  errorCode: null  
} 
    
export const logInReducer = (state: LogInReducer = initIdenitytReduerState, action: AuthAsyncActions): LogInReducer  => {
  switch (action.type) {
    case AUTH_ACTION_TYPES.LOG_IN_REQUEST:
      return {
        errorCode: null,
        authenticated: false,
        isFetching: true,
        didInvalidate: false,
      }
    case AUTH_ACTION_TYPES.LOG_IN_SUCCESS:
      return {
        isFetching: false,  
        didInvalidate: false,
        authenticated: true,
        errorCode: null 
      }
    case AUTH_ACTION_TYPES.LOG_IN_FAIL:
      return {
        authenticated: false,
        isFetching: false,
        didInvalidate: true,
        errorCode: action.errorCode
      }
    default: 
      return state
    }
  }