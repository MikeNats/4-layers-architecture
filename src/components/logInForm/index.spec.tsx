import React from 'react'
import LogInForm from '.'
import { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import {
  Switch,
  HashRouter as Router
} from 'react-router-dom';
import { auth } from '../../services/auth'
import {AppContext, initialAppContext } from '../../context'

jest.mock('../../services/auth');

Enzyme.configure({ adapter: new Adapter() })

   
describe("LogInForm", () => {
  describe('component', () => {
    let component
    it('renders as expected', () => {
      component = shallow(<LogInForm /> );

      expect(component).toMatchSnapshot();
    });
    it('should render a form', () => {	
		  component = mount(<Router><LogInForm /></Router>).find('form')
 
      expect(component.length).toBe(1)
  	});
    it('should render three fieldsets', () => {	
      component = mount(<Router><LogInForm /></Router>).find('fieldset')
      
      expect(component.length).toBe(3)
    });
    it('should render two lebels', () => {	
      component = mount(<Router><LogInForm /></Router>).find('label')

      expect(component.length).toBe(2)
      expect(component.at(0).text()).toBe('user name:')
      expect(component.at(1).text()).toBe('password:')
    });
    it('should render two inputs', () => {	
      component = mount(<Router><LogInForm /></Router>).find('input')

      expect(component.length).toBe(2)
    });
    it('should render one button', () => {	
      component = mount(<Router><LogInForm /></Router>).find('button')

      expect(component.length).toBe(1) 
      expect(component.text()).toEqual('submit')
    });
    it('should authenticate on submition and redirect', async () => {
      auth.mockImplementation(() => Promise.resolve({userId: 'sdf'}));
      const tick = () => new Promise(resolve =>
        setTimeout(resolve, 0))

      component = mount(<AppContext.Provider value={initialAppContext}>
        <Router><Switch><LogInForm /></Switch></Router></AppContext.Provider>)
      component.find('input').at(0).simulate('change', { currentTargert: { value: 'username' } });
      component.find('input').at(1).simulate('change', { currentTargert: { value: 'pass' } });
      component.find('button').simulate('click')
      await tick();

      component.update();
      
      expect(auth).toBeCalled()
      expect(initialAppContext.authenticated).toBe(true) 
      expect(component.find(LogInForm).props().location.pathname).toEqual('/home') 
      component.unmount()
    });
  });
});
