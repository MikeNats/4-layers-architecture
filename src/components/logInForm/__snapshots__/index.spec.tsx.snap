// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInForm component renders as expected 1`] = `ShallowWrapper {}`;
